File: src/main/java/com/nido/api/space/domain/model/SpaceRole.java
package com.nido.api.space.domain.model;

public enum SpaceRole {
    OWNER, ADMIN, MEMBER, VIEWER;

    /** Rang croissant en pouvoir : VIEWER < MEMBER < ADMIN < OWNER. */
    public int rank() {
        return switch (this) {
            case VIEWER -> 0;
            case MEMBER -> 1;
            case ADMIN -> 2;
            case OWNER -> 3;
        };
    }

    public boolean atLeast(SpaceRole required) {
        return rank() >= required.rank();
    }

    /** VIEWER est en lecture seule : toute écriture métier exige au moins MEMBER. */
    public boolean canWrite() {
        return this != VIEWER;
    }

    /** Gestion des membres, des invitations et de l'identité du groupe. */
    public boolean canManageSpace() {
        return this == OWNER || this == ADMIN;
    }
}

File: src/main/frontend/src/entities/space/lib/spaceRole.ts
import type { SpaceRole } from '../model/types'

const RANKS: Record<SpaceRole, number> = { VIEWER: 0, MEMBER: 1, ADMIN: 2, OWNER: 3 }

export function rank(role: SpaceRole): number {
  return RANKS[role]
}

/** Gérer les membres, les invitations et l'identité du groupe. */
export function canManageSpace(role: SpaceRole): boolean {
  return role === 'OWNER' || role === 'ADMIN'
}

/** Écrire du contenu métier. Le VIEWER est en lecture seule. */
export function canWrite(role: SpaceRole): boolean {
  return role !== 'VIEWER'
}

export function isOwner(role: SpaceRole): boolean {
  return role === 'OWNER'
}
